### PR TITLE
Update Callers to handle when getBlockType return undefined

### PR DESCRIPTION
**Pull request by amustaque97**
## Description

In most cases we either need to check if blockType is truthy before using it, or use optional chaining if it makes sense. eg ```const icon = blockType?.icon;```

## How has this been tested?

Testing Instruction can be found here #34346 (comment)
## Types of changes

Fix
## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 

cc @gwwar

### CODE REVIEW

File: packages/block-editor/src/components/copy-handler/index.js [hunk 1]
```diff
@@ -31,7 +31,7 @@ export function useNotifyCopy() {
 		let notice = '';
 		if ( selectedBlockClientIds.length === 1 ) {
 			const clientId = selectedBlockClientIds[ 0 ];
-			const { title } = getBlockType( getBlockName( clientId ) );
+			const { title } = getBlockType( getBlockName( clientId ) ) || null;
```
Comment (gwwar) [inline packages/block-editor/src/components/copy-handler/index.js]: Destructuring null will still cause an error. How about:
```
const title = getBlockType( getBlockName( clientId ) )?.title;
```

CleanShot 2021-09-23 at 13 06 19@2x

File: packages/block-editor/src/components/block-tools/block-selection-button.js [hunk 2]
```diff
@@ -76,7 +76,7 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
 	const { setNavigationMode, removeBlock } = useDispatch( blockEditorStore );
 	const ref = useRef();
 
-	const blockType = getBlockType( name );
+	const blockType = getBlockType( name ) || 'core/missing';
```
Comment (gwwar) [inline packages/block-editor/src/components/block-tools/block-selection-button.js]: Let's not use 'core/missing' directly here, since those internal details may change. One option here would be to drop this change, but update `getAccessibleBlockLabel` to not error when blockType is undefined.

File: packages/blocks/src/api/raw-handling/shortcode-converter.js [hunk 3]
```diff
@@ -92,8 +92,15 @@ function segmentHTMLToShortcodeBlock(
 			( schema ) => schema.shortcode( match.shortcode.attrs, match )
 		);
 
+		const blockType = getBlockType( transformation.blockName );
+		if ( ! blockType ) {
+			return [
+				...segmentHTMLToShortcodeBlock( beforeHTML ),
+				...segmentHTMLToShortcodeBlock( afterHTML ),
+			];
+		}
```
Comment (gwwar) [inline packages/blocks/src/api/raw-handling/shortcode-converter.js]: Do you mind moving this change into a separate PR, I think we'll want a unit test here to make sure we're handling this case properly.

**Review by gwwar — COMMENTED** [review]
Thanks for taking a look @amustaque97! I left a few notes.

File: packages/block-editor/src/components/copy-handler/index.js (hunk 1 above)
Comment (amustaque97) [reply packages/block-editor/src/components/copy-handler/index.js]: @gwwar, my bad. I didn't notice this change. This is not my final change. Because I tested this thing for other places. My dev environment (Github codespaces) stopped working abruptly 😩  Not sure why! Few changes might get pushed which were not part of the final changes. Let me fix it.

File: packages/block-editor/src/components/block-tools/block-selection-button.js (hunk 2 above)
Comment (amustaque97) [reply packages/block-editor/src/components/block-tools/block-selection-button.js]: Again, not sure. I remember ```getBlockType``` returns an object with properties like ``` title, attributes, icon``` passing string `core/missing` is not correct. Let me fix this as well.

File: packages/blocks/src/api/raw-handling/shortcode-converter.js (hunk 3 above)
Comment (amustaque97) [reply packages/blocks/src/api/raw-handling/shortcode-converter.js]: Ok sure, I will remove these changes from this PR and will create a separate PR. It might take some time for me to come up with the PR. I will be writing front-end/javascript test cases for the first time. Need some upskilling 😄

File: packages/block-editor/src/components/block-tools/block-selection-button.js (hunk 2 above)
Comment (amustaque97) [reply packages/block-editor/src/components/block-tools/block-selection-button.js]: @gwwar, I have changed ```getAccessibleBlockLabel``` function definition. What do you think about this change? Looking forward to your inputs before pushing this change.
``` javascript
export function getAccessibleBlockLabel(
	blockType,
	attributes,
	position,
	direction = 'vertical'
) {
	// `title` is already localized, `label` is a user-supplied value.
	const title = blockType?.title;
	const label = blockType ? getBlockLabel( blockType, attributes, 'accessibility' ) : '';
	const hasPosition = position !== undefined;

	// getBlockLabel returns the block title as a fallback when there's no label,
	// if it did return the title, this function needs to avoid adding the
	// title twice within the accessible label. Use this `hasLabel` boolean to
	// handle that.
	const hasLabel = label && label !== title;

	if ( hasPosition && direction === 'vertical' ) {
		if ( hasLabel ) {
			return sprintf(
				/* translators: accessibility text. 1: The block title. 2: The block row number. 3: The block label.. */
				__( '%1$s Block. Row %2$d. %3$s' ),
				title,
				position,
				label
			);
		}

		return sprintf(
			/* translators: accessibility text. 1: The block title. 2: The block row number. */
			__( '%1$s Block. Row %2$d' ),
			title,
			position
		);
	} else if ( hasPosition && direction === 'horizontal' ) {
		if ( hasLabel ) {
			return sprintf(
				/* translators: accessibility text. 1: The block title. 2: The block column number. 3: The block label.. */
				__( '%1$s Block. Column %2$d. %3$s' ),
				title,
				position,
				label
			);
		}

		return sprintf(
			/* translators: accessibility text. 1: The block title. 2: The block column number. */
			__( '%1$s Block. Column %2$d' ),
			title,
			position
		);
	}

	if ( hasLabel ) {
		return sprintf(
			/* translators: accessibility text. %1: The block title. %2: The block label. */
			__( '%1$s Block. %2$s' ),
			title,
			label
		);
	}

	return sprintf(
		/* translators: accessibility text. %s: The block title. */
		__( '%s Block' ),
		title
	);
}
```

**Comment by gwwar** [comment]
This PR is part of https://github.com/WordPress/gutenberg/issues/34462

To manually test this, create a new post, and in the code editor add:
```
<!-- wp:foo /-->
```

Go back to code view, and we should see the missing block:
Screen Shot 2021-08-31 at 3 18 22 PM

Unregister the missing block type in console:
```
wp.data.dispatch('core/blocks').removeBlockTypes('core/missing')
```

Try doing some actions like going to code view or trying to click the missing block. Theoretically the editor should not WSOD, or throw a JS error (from any of the areas this PR has updated).

File: packages/block-editor/src/components/copy-handler/index.js [hunk 4]
```diff
@@ -31,7 +31,7 @@ export function useNotifyCopy() {
 		let notice = '';
 		if ( selectedBlockClientIds.length === 1 ) {
 			const clientId = selectedBlockClientIds[ 0 ];
-			const { title } = getBlockType( getBlockName( clientId ) );
+			const { title } = getBlockType( getBlockName( clientId ) )?.title;
```
Comment (gwwar) [inline packages/block-editor/src/components/copy-handler/index.js]: This one needs to be updated to:
`const title = getBlockType( getBlockName( clientId ) )?.title;`

**Review by gwwar — COMMENTED** [review]
This one just needs one more change and it should be good to land! Sorry for the delay @amustaque97 

I also verified in manual testing that we didn't see any errors related to code changes here when unregistering the missing blockType

**Comment by amustaque97** [comment]
@gwwar, I have addressed review comments. Please take a look once you get some time.

**Review by gwwar — APPROVED** [review]
Thanks for your work here @amustaque97! I'll land this once we get a clean test run ✅
